### PR TITLE
Parse any/all challenges served in WWW-Authenticate headers

### LINKED ISSUE #992: Authorization header fails to parse authenticate challenges where Bearer is not the first challenge
Android version of [`azure-activedirectory-library-for-objc/#1042`](https://github.com/AzureAD/azure-activedirectory-library-for-objc/issues/1042)

>If a WWW-Authenticate header has a value like `Basic realm="myrealm", Bearer ...` then we fail to parse it correctly. <br/>
OAuth 2.0 Authorization Protocol Spec: http://self-issued.info/docs/draft-ietf-oauth-v2-bearer-16.html
HTTP Authenticate Spec: https://tools.ietf.org/html/draft-ietf-httpbis-p7-auth-18#page-12
HTTP Message Parsing Spec: https://tools.ietf.org/html/draft-ietf-httpbis-p1-messaging-18#section-1.2.2

### PR DESCRIPTION
**Pull request by iambmelt**
This closes #992 

These changes add support for parsing any/all challenge schemes served by `WWW-Authenticate` headers. It also changes the `Exception` message to be more accurate, in instances where parsing fails due to absent or malformed `authorization_uri` parameters.

New tests have been added to satisfy the problem description in #992

### CODE REVIEW

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 1]
```diff
@@ -195,80 +192,357 @@ public void run() {
...
-                        }
-
-                        headerItems.put(key, value);
-                    } else {
-                        // invalid format
-                        throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
-                    }
+        }
+
+        //testing...
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Clean up this comment

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 2]
```diff
@@ -195,80 +192,356 @@ public void run() {
...
+                if (params.containsKey(key)) {
+                    Logger.w(TAG,
+                            "Key/value pair list contains redundant key. ",
+                            "Redundant key: " + key,
+                            ADALError.DEVELOPER_BEARER_HEADER_MULTIPLE_ITEMS);
                 }
-            } else {
+
+                // Add the key/value to the Map
+                params.put(key, value);
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Investigate: Do these keys/values need to be passed through `StringExtensions.urlFormDecode(String)` ?

Comment (iambmelt) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: My testing is showing that the in/out values are not changed - the existing unit tests don't account for this. Will leave as-is for now and be ready to adjust if partner testing yields any issues

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 3]
```diff
@@ -195,80 +192,356 @@ public void run() {
...
+     * @see <a href-"https://tools.ietf.org/html/rfc7617">RFC-7617</a>
+     * @see <a href="https://tools.ietf.org/html/rfc6750">RFC-6750</a>
+     */
+    private static class Challenge {
+
+        /**
+         * Regex sequence intended to be prefixed with another value. Whichever value precedes it
+         * will be parsed/grouped-out, assuming it is not bounded by a par of double-quotes ("").
+         */
+        private static final String REGEX_UNQUOTED_LOOKAHEAD = "(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)";
```
Comment (heidijinxujia) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Pardon me if I miss some context here. So we are changing the entire regex here? The regex here is different from the dev branch. What will happen if the challenge begin with schema "Bearer"? By calling parseScheme, right?

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 4]
```diff
@@ -195,80 +192,356 @@ public void run() {
...
+
+        /**
+         * Parses the scheme of a challenge String.
+         *
+         * @param challenge The challenge String to parse.
+         * @return The scheme portion of the challenge String.
+         * @throws ResourceAuthenticationChallengeException If a parsing error is encountered or
+         *                                                  the String is malformed.
+         */
+        private static String parseScheme(String challenge) throws ResourceAuthenticationChallengeException {
```
Comment (heidijinxujia) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: So now it will accept all kinds of scheme, right? Not just "BEARER"?

Comment (iambmelt) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Correct, any scheme can now be used/parsed

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java (hunk 3 above)
Comment (iambmelt) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: The previous regex required that the `WWW-Authenticate` header begin with `Bearer` and not contain other challenges. The new parser should be tolerant of an arbitrary number of challenges and parse them. The `Bearer` challenge is only sought at the end to return the correct `AuthenticationParameters` object

**Comment by iambmelt** [comment]
Manual test step.

Use any HTTP requesting application/service to

`GET https://graph.microsoft.com/v1.0/me/`

This should present a 401 challenge

Headers
```
client-request-id → <omitted>
content-length → 241
content-type → application/json; charset=utf-8
date → Sat, 06 Jan 2018 00:16:36 GMT
request-id → 3371374e-1f5f-47dd-b30c-8b75f4681037
www-authenticate → Bearer realm="", authorization_uri="https://login.microsoftonline.com/common/oauth2/authorize", client_id="00000003-0000-0000-c000-000000000000"
x-ms-ags-diagnostic →{"ServerInfo":{"DataCenter":"East US","Slice":"SliceA","Ring":"5","ScaleUnit":"001","Host":"AGSFE_IN_23","ADSiteName":"EST"}}
```

Response body
```json
{
  "error": {
    "code": "InvalidAuthenticationToken",
    "message": "Bearer access token is empty.",
    "innerError": {
      "request-id": "09c66fa0-34e0-4557-917c-1d570718944e",
      "date": "2018-01-06T00:26:37"
    }
  }
}
```

Which, when parsed yields the following `AuthenticationParameters` object:

```java
AuthenticationParameters [mAuthority="https://login.microsoftonline.com/common/oauth2/authorize", mResource=""] 
```

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 5]
```diff
@@ -195,80 +192,356 @@ public void run() {
...
+            // Split on unquoted commas
+            final Map<String, String> params = new HashMap<>();
+            final String[] splitOnUnquotedCommas = challengeSansScheme.split(REGEX_SPLIT_UNQUOTED_COMMA, -1);
+            for (final String paramSet : splitOnUnquotedCommas) {
+                // Split keys/values by the '='
+                final String[] splitOnUnquotedEquals = paramSet.split(REGEX_SPLIT_UNQUOTED_EQUALS, -1);
+
+                // We should now have a left-side and right-side
+                if (splitOnUnquotedEquals.length != 2) {
+                    // Is this really what you want?
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Clean up this comment...

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 6]
```diff
@@ -195,80 +192,356 @@ public void run() {
...
+                    throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
+                    //continue; // If there's no key/value pair, skip this token
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Remove commented code

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 7]
```diff
@@ -195,80 +192,356 @@ public void run() {
...
 
-                String authority = headerItems.get(AUTHORITY_KEY);
-                if (!StringExtensions.isNullOrBlank(authority)) {
-                    authParams = new AuthenticationParameters(
-                            StringExtensions.removeQuoteInHeaderValue(authority),
-                            StringExtensions.removeQuoteInHeaderValue(headerItems.get(RESOURCE_KEY)));
-                } else {
-                    // invalid format
-                    throw new ResourceAuthenticationChallengeException(AUTH_HEADER_MISSING_AUTHORITY);
+                // Create the keys/values, trimming off any bogus whitespace
```
Comment (iambmelt) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Change to 'excess'

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 8]
```diff
@@ -195,80 +192,354 @@ public void run() {
...
+        /**
+         * Parses a single challenge String, typically the value field of a WWW-Authenticate header.
+         *
+         * @param challenge The challenge String to parse.
+         * @return The Challenge object derived from challenge.
+         * @throws ResourceAuthenticationChallengeException If a parsing error is encountered or
+         *                                                  the String is malformed.
+         */
+        static Challenge parseChallenge(final String challenge) throws ResourceAuthenticationChallengeException {
+            final String scheme = parseScheme(challenge);
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: do we need null checks for scheme, challengeSansScheme and challenge?

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 9]
```diff
@@ -195,80 +192,354 @@ public void run() {
...
+         */
+        private static List<String> separateChallenges(final String challenges) {
+            // Split the supplied String on those commas which are not constrained by quotes
+            String[] splitOnUnquotedCommas = challenges.split(REGEX_SPLIT_UNQUOTED_COMMA, -1);
+            sanitizeWhitespace(splitOnUnquotedCommas);
+            List<String> tokensContainingScheme = extractTokensContainingScheme(splitOnUnquotedCommas);
+
+            // init an array to store the out-values
+            String[] outStrings = new String[tokensContainingScheme.size()];
+            for (int ii = 0; ii < outStrings.length; ii++) {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: just curious: ii vs i vs index?

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java [hunk 10]
```diff
@@ -195,80 +192,354 @@ public void run() {
...
-                    }
+        }
+
+        final List<Challenge> challenges = Challenge.parseChallenges(authenticateHeader);
+
+        // Grab the Bearer challenge
+        Challenge bearerChallenge = null;
+
+        for (final Challenge challenge : challenges) {
+            if (challenge.getScheme().equalsIgnoreCase(BEARER)) {
```
Comment (addev-ashish) [inline adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: you are assuming getScheme can never return null. I would flip the check around

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java (hunk 9 above)
Comment (iambmelt) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Yep -- I tend to favor using `ii` for indexes as `i` isn't very `ctrl-f` searchable. I'm fine with `index` too, mostly `ii` is a matter of habit/style at this point

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java (hunk 10 above)
Comment (iambmelt) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Resolved in 78b52908d59b9e8fbac42ce896314e58063224da

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java (hunk 8 above)
Comment (iambmelt) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: Resolved in 78b52908d59b9e8fbac42ce896314e58063224da

**Comment by iambmelt** [comment]
Added additional logging/`null` checking in 78b52908d59b9e8fbac42ce896314e58063224da

/cc @addev-ashish @heidijinxujia

**Comment by heidijinxujia** [comment]
LGTM

File: adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java (hunk 9 above)
Comment (addev-ashish) [reply adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java]: ok.

**Pull request merged**